### ISSUE TITLE
LoRA with query groups doesn't work correctly

### ISSUE THREAD

**Issue by Andrei-Aksionov** [Lightning-AI/lit-gpt#357]
This is a more technical analysis of this issue: https://github.com/Lightning-AI/lit-gpt/issues/348 (check it to the see the error output)

TL;DR: `groups` in `nn.Conv1d` splits tensor in an equal parts and they are processed independently, but if the number of query groups is smaller than the number of heads then in a combined matrix `QKV` key and value parts will be smaller than query one and thus will be processed incorrectly by `nn.Conv1d`. 

---

In `LoRAQKVLinear` we have a combined matrix QKV that consists of three matrices: query, key and value. So the output dimension of the matrix consists of three terms: number of heads * head size * 3 (because all three matrices are of equal size). 
But with [query groups](https://github.com/Lightning-AI/lit-gpt/blob/e40647fd7ed4a9b5c8634905546835757eb7c9b7/lit_gpt/config.py#L29) the actual dimension will be: 
1. query: `number of heads` * `head size`
2. key: `number of query groups` * `head_size`
3. value: `number of query groups` * `head_size`

So if the `number of query groups` < `number of heads`, then the portion of key and value in the combined matrix will be smaller than the portion of the query. In other words QKV will consist of parts of unequal size.
And the problem is in how `nn.Conv1d` works with groups parameter. For instance for groups=2 from the official documentation:
> At groups=2, the operation becomes equivalent to having two conv layers side by side, each seeing half the input channels and producing half the output channels, and both subsequently concatenated.

Which means that no matter what the number of groups is, the tensor will be divided in equal parts and each part is processed independently. But as I stated above the tensor might consist of unequal sized parts.

The issue was overlooked as the number of groups was 2 (`enable_lora` was hardcoded as (True, False, True)) and the dimension of [`after_A`](https://github.com/Lightning-AI/lit-gpt/blob/e40647fd7ed4a9b5c8634905546835757eb7c9b7/lit_gpt/lora.py#L370) is even thus divisible by two and thus no errors are raised but in fact it is incorrect. With the number of groups of 3 (LoRA is applied to all three matrices in QKV) we get the error.

---

Ideally would be to have an ability to provide a custom way to split tensor in `nn.Conv1d`, but it seems to be impossible.
So the only option is to do everything manually (if number of query groups != number of heads):
1. split `after_A` tensor
2. split `lora_B` weight matrix
3. apply independently
4. concatenate the result

But of course it's not particularly performant.
Any ideas?

**Comment by reeceshuttle** [comment]
I have noticed this exact same bug and agree completely with your analysis. (Was literally coming here to post this issue but you beat me to it 😂) I have directly observed that while there is no error with `enable_lora` set to (True, False, True), incorrect multiplications are done. 

My way around this has been to implement a separate A and B matrix for each qkv individually, which is how a standard LoRA layer is implemented. I have run this code and have seen that it does not have an error when `enable_lora` is set to (True, True, True) when running on Falcon 7B and have seen loss decrease. Also, I have analyzed the forward passes of my code and get the same result as the old code when the query groups are equal to the number of heads, which makes me feel confident that my code is correct.

I have yet to look into performance in detail, but runtime for a particular iter in fine-tuning appears to be around the same between my new code and the old code.

Thank you for writing up this analysis!

**Comment by JustinGoheen** [comment]
I wrote a tensor split in scripts/convert_lit_checkpoint that may help

https://github.com/Lightning-AI/lit-gpt/blob/f6aae7870728043ed4cf285721d21503997dc14c/scripts/convert_lit_checkpoint.py#L152
